{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from collections import Counter,defaultdict\n",
    "import networkx as nx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_csv('NAS_processed_032523.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clean up raw data and create network input dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "publications = data['Publication_Title']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "publications_list = []\n",
    "\n",
    "for v in publications:\n",
    "    try:\n",
    "        if float(v):\n",
    "            publications_list.append(temp)\n",
    "    except:\n",
    "        temp = v\n",
    "        publications_list.append(v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "publications_list_cap = []\n",
    "for v in publications_list:\n",
    "    first = v[0]\n",
    "    publications_list_cap.append(first.upper()+v[1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "Exposure_Layer2 = data['Exposure_Layer2']\n",
    "Outcome_Layer2 = data['Outcome_Layer2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "intersection = set(Exposure_Layer2).intersection(set(Outcome_Layer2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "hierach1 = defaultdict(set)\n",
    "edge2publication = defaultdict(list)\n",
    "for i,v in enumerate(Exposure_Layer2):\n",
    "    hierach1[v].add(Outcome_Layer2[i])\n",
    "    edge2publication[(v,Outcome_Layer2[i])].append(publications_list_cap[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "exposures_l2 = []\n",
    "outcomes_l2 = []\n",
    "exposure_cate = []\n",
    "outcome_cate = []\n",
    "edge_ID2pub = dict()\n",
    "i = 1\n",
    "for k in sorted(hierach1.keys()): ##031823\n",
    "    for v in sorted(hierach1[k]): ##\n",
    "        exposures_l2.append(k)\n",
    "        outcomes_l2.append(v)\n",
    "        edge_ID2pub[i] = edge2publication[(k,v)]\n",
    "        i+=1\n",
    "        if k not in intersection:\n",
    "            exposure_cate.append('exposures')\n",
    "        else:\n",
    "            exposure_cate.append('both')\n",
    "        if v not in intersection:\n",
    "            outcome_cate.append('outcomes')\n",
    "        else:\n",
    "            outcome_cate.append('both')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_id_c = max(edge_ID2pub.keys())+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_id = np.arange(1,edge_id_c).tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "layer2_dataframe = pd.DataFrame(list(zip(exposures_l2, outcomes_l2,edge_id,exposure_cate,outcome_cate)),columns =['Exposures', 'Outcomes','edge_id','exposure_attr','outcome_attr'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Exposures</th>\n",
       "      <th>Outcomes</th>\n",
       "      <th>edge_id</th>\n",
       "      <th>exposure_attr</th>\n",
       "      <th>outcome_attr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BMI</td>\n",
       "      <td>cholesterol level</td>\n",
       "      <td>1</td>\n",
       "      <td>exposures</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DNAmethylation</td>\n",
       "      <td>blood pressure</td>\n",
       "      <td>2</td>\n",
       "      <td>both</td>\n",
       "      <td>outcomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DNAmethylation</td>\n",
       "      <td>incident cancer</td>\n",
       "      <td>3</td>\n",
       "      <td>both</td>\n",
       "      <td>outcomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DNAmethylation</td>\n",
       "      <td>leukocyte telomere length</td>\n",
       "      <td>4</td>\n",
       "      <td>both</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DNAmethylation</td>\n",
       "      <td>lung function</td>\n",
       "      <td>5</td>\n",
       "      <td>both</td>\n",
       "      <td>outcomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>tocopherol levels</td>\n",
       "      <td>lung function</td>\n",
       "      <td>126</td>\n",
       "      <td>exposures</td>\n",
       "      <td>outcomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>toenail metals</td>\n",
       "      <td>DNAmethylation</td>\n",
       "      <td>127</td>\n",
       "      <td>both</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>toenail metals</td>\n",
       "      <td>blood pressure</td>\n",
       "      <td>128</td>\n",
       "      <td>both</td>\n",
       "      <td>outcomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>toenail metals</td>\n",
       "      <td>metabolites</td>\n",
       "      <td>129</td>\n",
       "      <td>both</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>urinary metals</td>\n",
       "      <td>DNAmethylation</td>\n",
       "      <td>130</td>\n",
       "      <td>both</td>\n",
       "      <td>both</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>130 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Exposures                   Outcomes  edge_id exposure_attr  \\\n",
       "0                  BMI          cholesterol level        1     exposures   \n",
       "1       DNAmethylation             blood pressure        2          both   \n",
       "2       DNAmethylation            incident cancer        3          both   \n",
       "3       DNAmethylation  leukocyte telomere length        4          both   \n",
       "4       DNAmethylation              lung function        5          both   \n",
       "..                 ...                        ...      ...           ...   \n",
       "125  tocopherol levels              lung function      126     exposures   \n",
       "126     toenail metals             DNAmethylation      127          both   \n",
       "127     toenail metals             blood pressure      128          both   \n",
       "128     toenail metals                metabolites      129          both   \n",
       "129     urinary metals             DNAmethylation      130          both   \n",
       "\n",
       "    outcome_attr  \n",
       "0           both  \n",
       "1       outcomes  \n",
       "2       outcomes  \n",
       "3           both  \n",
       "4       outcomes  \n",
       "..           ...  \n",
       "125     outcomes  \n",
       "126         both  \n",
       "127     outcomes  \n",
       "128         both  \n",
       "129         both  \n",
       "\n",
       "[130 rows x 5 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "layer2_dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame.to_csv(layer2_dataframe,'processed_graph_layer2.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create edge number to publication list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "edg_list = []\n",
    "pub_list = []\n",
    "for k in range(1,edge_id_c):\n",
    "    for v in set(edge_ID2pub[k]):\n",
    "        edg_list.append(k)\n",
    "        pub_list.append(v)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "layer2_publist = pd.DataFrame(list(zip(edg_list, pub_list)),columns =['Edge Id', 'Publication'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Edge Id</th>\n",
       "      <th>Publication</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Lifestyle change and high-density lipoprotein ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>Association between blood pressure and DNA met...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>Longitudinal study of DNA methylation of infla...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>Mirna-processing gene methylation and cancer risk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>Promoter methylation of pgc1a and pgc1b predic...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>126</td>\n",
       "      <td>Serum tocopherol levels and vitamin e intake a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>127</td>\n",
       "      <td>Arsenic exposure and DNA methylation among eld...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>128</td>\n",
       "      <td>Associations of toenail arsenic, cadmium, merc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>129</td>\n",
       "      <td>Metabolomic signatures of lead exposure in the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>130</td>\n",
       "      <td>Individual species and cumulative mixture rela...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>172 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Edge Id                                        Publication\n",
       "0          1  Lifestyle change and high-density lipoprotein ...\n",
       "1          2  Association between blood pressure and DNA met...\n",
       "2          3  Longitudinal study of DNA methylation of infla...\n",
       "3          3  Mirna-processing gene methylation and cancer risk\n",
       "4          3  Promoter methylation of pgc1a and pgc1b predic...\n",
       "..       ...                                                ...\n",
       "167      126  Serum tocopherol levels and vitamin e intake a...\n",
       "168      127  Arsenic exposure and DNA methylation among eld...\n",
       "169      128  Associations of toenail arsenic, cadmium, merc...\n",
       "170      129  Metabolomic signatures of lead exposure in the...\n",
       "171      130  Individual species and cumulative mixture rela...\n",
       "\n",
       "[172 rows x 2 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "layer2_publist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame.to_csv(layer2_publist,'layer2_publist.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Layered node keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "Exposure_Layer1 = data['Exposure_Layer1']\n",
    "Outcome_Layer1 = data['Outcome_Layer1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "Exposure_Layer2 = data['Exposure_Layer2']\n",
    "Outcome_Layer2 = data['Outcome_Layer2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "Exposure_Layer3 = data['Exposure_Layer3']\n",
    "Outcome_Layer3 = data['Outcome_Layer3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "chemical\n",
      "air pollution\n",
      "\n",
      "\n",
      "physical\n",
      "relative humidity\n",
      "solar and geomagnetic activity\n",
      "occupation\n",
      "temperature\n",
      "road proximity\n",
      "\n",
      "\n",
      "biomarker\n",
      "genes\n",
      "biological aging\n",
      "metabolites\n",
      "cholesterol level\n",
      "metabolic syndrome component\n",
      "leukocyte telomere length\n",
      "urinary metals\n",
      "BMI\n",
      "mitochondrial DNA\n",
      "DNAmethylation\n",
      "blood lead\n",
      "bone lead\n",
      "hormone\n",
      "tocopherol levels\n",
      "toenail metals\n",
      "\n",
      "\n",
      "psychosocial\n",
      "anxiety\n",
      "depression\n",
      "optimism\n",
      "family\n",
      "happiness\n",
      "hassles and uplifts\n",
      "psychosocial stressors\n",
      "life satisfaction\n",
      "openness to experiences\n",
      "self-regulation\n",
      "emotional reactivity\n",
      "pessimism\n",
      "co-occurrence of affect\n",
      "stressful life events\n",
      "hostility\n",
      "\n",
      "\n",
      "behavior\n",
      "alcohol\n",
      "diet\n",
      "smoking\n",
      "sleep\n",
      "drug use\n",
      "\n",
      "\n",
      "disease\n",
      "asthma\n",
      "allergic phenotypes\n",
      "hypertension\n",
      "diabetes\n"
     ]
    }
   ],
   "source": [
    "#Exposure Layer1-Layer2 hierachy\n",
    "hierach1=defaultdict(set)\n",
    "for i,v in enumerate(Exposure_Layer1):\n",
    "    hierach1[v].add(Exposure_Layer2[i])\n",
    "\n",
    "for k in hierach1.keys():\n",
    "    print('\\n')\n",
    "    print(k)\n",
    "    for v in hierach1[k]:\n",
    "        #print('\\t'+v)\n",
    "        print(v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "air pollution\n",
      "long-term PM2.5 elemental components\n",
      "intermediate-term CO\n",
      "long-term ozone\n",
      "short-term NO\n",
      "short-term PM2.5 elemental components\n",
      "intermediate-term black carbon\n",
      "intermediate-term NO\n",
      "long-term nitrate\n",
      "short-term CO\n",
      "long-term NO2\n",
      "short-term SO2\n",
      "long-term ammonium\n",
      "intermediate-term PM2.5 mass\n",
      "intermediate-term nitrate\n",
      "long-term sulfate\n",
      "short-term NO2\n",
      "short-term organic carbon\n",
      "short-term nitrate\n",
      "short-term sulfate\n",
      "long-term PM2.5 mass\n",
      "short-term black carbon\n",
      "intermediate-term PM2.5 elemental components\n",
      "intermediate-term organic carbon\n",
      "long-term black carbon\n",
      "long-term organic carbon\n",
      "particle radioactivity\n",
      "intermediate-term particle number\n",
      "intermediate-term NO2\n",
      "short-term PM2.5 mass\n",
      "intermediate-term sulfate\n",
      "short-term particle number\n",
      "short-term ozone\n",
      "intermediate-term ozone\n",
      "\n",
      "\n",
      "solar and geomagnetic activity\n",
      "interplanetary magnetic field\n",
      "sunspot number\n",
      "geomagnetic activity-K index\n",
      "\n",
      "\n",
      "temperature\n",
      "short-term temperature standardad deviation\n",
      "short-term temperature\n",
      "long-term temperature\n",
      "short-term temperature-indoor\n",
      "spatial and temporal temperature variation\n",
      "long-term temperature-outdoor\n",
      "intermediate-term temperature standardad deviation\n",
      "short-term temperature-outdoor\n",
      "intermediate-term temperature\n",
      "\n",
      "\n",
      "metabolites\n",
      "one-carbon metabolites\n",
      "\n",
      "\n",
      "bone lead\n",
      "tibia lead\n",
      "patella lead\n",
      "\n",
      "\n",
      "mitochondrial DNA\n",
      "mitochondrial DNA copy number\n",
      "\n",
      "\n",
      "blood lead\n",
      "blood lead\n",
      "\n",
      "\n",
      "toenail metals\n",
      "toenail arsenic\n",
      "toenail mercury\n",
      "toenail cadmium\n",
      "toenail manganese\n",
      "toenail lead\n",
      "\n",
      "\n",
      "DNAmethylation\n",
      "genes-450K\n",
      "genes-selected\n",
      "aging clocks\n",
      "\n",
      "\n",
      "leukocyte telomere length\n",
      "leukocyte telomere length\n",
      "\n",
      "\n",
      "urinary metals\n",
      "urinary manganese\n",
      "urinary mercury\n",
      "urinary cadmium\n",
      "urinary arsenic\n",
      "urinary lead\n",
      "\n",
      "\n",
      "psychosocial stressors\n",
      "psychosocial stressors\n",
      "\n",
      "\n",
      "family\n",
      "early life experience-cherished/harshly disciplined/ordinary\n",
      "childhood socioeconomic status\n",
      "close relationships\n",
      "\n",
      "\n",
      "optimism\n",
      "optimism\n",
      "\n",
      "\n",
      "smoking\n",
      "smoking\n",
      "\n",
      "\n",
      "drug use\n",
      "antihypertensive medication\n",
      "\n",
      "\n",
      "diet\n",
      "western diet\n",
      "vitamin E intake\n",
      "fiber content\n",
      "prudent diet \n",
      "dietary flavonoids\n",
      "manganese intake\n",
      "\n",
      "\n",
      "road proximity\n",
      "road proximity\n",
      "\n",
      "\n",
      "genes\n",
      "HFE gene\n",
      "APOE genetic variants\n",
      "folate-related genes\n",
      "transferrin gene\n",
      "genes-SNPs\n",
      "\n",
      "\n",
      "relative humidity\n",
      "intermediate-term relative humidity\n",
      "short-term relative humidity\n",
      "\n",
      "\n",
      "hormone\n",
      "irisin\n",
      "chemerin\n",
      "leptin\n",
      "\n",
      "\n",
      "diabetes\n",
      "diabetes\n",
      "\n",
      "\n",
      "depression\n",
      "depression\n",
      "\n",
      "\n",
      "anxiety\n",
      "anxiety\n",
      "\n",
      "\n",
      "hostility\n",
      "hostility\n",
      "\n",
      "\n",
      "happiness\n",
      "happiness\n",
      "\n",
      "\n",
      "life satisfaction\n",
      "life satisfaction\n",
      "\n",
      "\n",
      "hassles and uplifts\n",
      "hassles and uplifts\n",
      "\n",
      "\n",
      "co-occurrence of affect\n",
      "co-occurrence of affect\n",
      "\n",
      "\n",
      "tocopherol levels\n",
      "serum tocopherol levels\n",
      "\n",
      "\n",
      "cholesterol level\n",
      "high-density lipoprotein cholesterol\n",
      "total cholesterol\n",
      "\n",
      "\n",
      "emotional reactivity\n",
      "emotional reactivity\n",
      "\n",
      "\n",
      "occupation\n",
      "job categories\n",
      "\n",
      "\n",
      "pessimism\n",
      "pessimism\n",
      "\n",
      "\n",
      "sleep\n",
      "sleep\n",
      "\n",
      "\n",
      "hypertension\n",
      "incidence hypertension\n",
      "duration of hypertension\n",
      "\n",
      "\n",
      "biological aging\n",
      "biological aging\n",
      "\n",
      "\n",
      "asthma\n",
      "asthma\n",
      "\n",
      "\n",
      "allergic phenotypes\n",
      "allergic phenotypes\n",
      "\n",
      "\n",
      "alcohol\n",
      "alcohol\n",
      "\n",
      "\n",
      "BMI\n",
      "BMI\n",
      "\n",
      "\n",
      "openness to experiences\n",
      "openness to experiences\n",
      "\n",
      "\n",
      "metabolic syndrome component\n",
      "metabolic syndrome component\n",
      "\n",
      "\n",
      "stressful life events\n",
      "stressful life events\n",
      "\n",
      "\n",
      "self-regulation\n",
      "self-regulation\n"
     ]
    }
   ],
   "source": [
    "#Exposure Layer2-Layer3 hierachy\n",
    "hierach2 = defaultdict(set)\n",
    "for i,v in enumerate(Exposure_Layer2):\n",
    "    hierach2[v].add(Exposure_Layer3[i])\n",
    "\n",
    "for k in hierach2.keys():\n",
    "    print('\\n')\n",
    "    print(k)\n",
    "    for v in hierach2[k]:\n",
    "        #print('\\t'+v)\n",
    "        print(v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "intermediate-term temperature\n",
      "intermediate-term temperature standardad deviation\n",
      "long-term temperature\n",
      "long-term temperature-outdoor\n",
      "short-term temperature\n",
      "short-term temperature standardad deviation\n",
      "short-term temperature-indoor\n",
      "short-term temperature-outdoor\n",
      "spatial and temporal temperature variation\n"
     ]
    }
   ],
   "source": [
    "for s in sorted(hierach2['temperature']):\n",
    "    print(s)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "intermediate-term CO\n",
      "intermediate-term NO\n",
      "intermediate-term NO2\n",
      "intermediate-term PM2.5 elemental components\n",
      "intermediate-term PM2.5 mass\n",
      "intermediate-term black carbon\n",
      "intermediate-term nitrate\n",
      "intermediate-term organic carbon\n",
      "intermediate-term ozone\n",
      "intermediate-term particle number\n",
      "intermediate-term sulfate\n",
      "long-term NO2\n",
      "long-term PM2.5 elemental components\n",
      "long-term PM2.5 mass\n",
      "long-term ammonium\n",
      "long-term black carbon\n",
      "long-term nitrate\n",
      "long-term organic carbon\n",
      "long-term ozone\n",
      "long-term sulfate\n",
      "particle radioactivity\n",
      "short-term CO\n",
      "short-term NO\n",
      "short-term NO2\n",
      "short-term PM2.5 elemental components\n",
      "short-term PM2.5 mass\n",
      "short-term SO2\n",
      "short-term black carbon\n",
      "short-term nitrate\n",
      "short-term organic carbon\n",
      "short-term ozone\n",
      "short-term particle number\n",
      "short-term sulfate\n"
     ]
    }
   ],
   "source": [
    "for s in sorted(hierach2['air pollution']):\n",
    "    print(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "biomarker\n",
      "incident abnormal cardiac conductivity\n",
      "incident metabolic syndrome\n",
      "blood urea nitrogen\n",
      "endothelial dysfunction markers\n",
      "lipid profile\n",
      "serum uric acid\n",
      "arterial stiffness\n",
      "neural plasticity\n",
      "cholesterol level\n",
      "leukocyte telomere length change rate\n",
      "olfactory recognition\n",
      "oxidative DNA damage\n",
      "inflammation biomarkers\n",
      "fine motor function\n",
      "urinary metals\n",
      "plasma homocysteine\n",
      "fasting blood glucose\n",
      "alzheimer disease biomarkers\n",
      "white blood cell counts\n",
      "metabolic syndrome components\n",
      "toenail metals\n",
      "glomerular filtration rate\n",
      "metabolites\n",
      "microRNA\n",
      "leukocyte telomere length\n",
      "leukocyte distribution\n",
      "intraocular pressure\n",
      "mitochondrial DNA\n",
      "cognitive function\n",
      "DNAmethylation\n",
      "intima-media thickness of the common carotid artery\n",
      "blood pressure\n",
      "electrocardiogram\n",
      "hemoglobin concentration\n",
      "blood lead\n",
      "bone lead\n",
      "lung function\n",
      "\n",
      "\n",
      "disease\n",
      "cardiovascular disease\n",
      "tooth loss\n",
      "tremor\n",
      "mortality\n",
      "glaucoma\n",
      "resistant hypertension\n",
      "incident cancer\n",
      "periodontal disease progression\n",
      "incident chronic kidney disease\n",
      "diabetes\n",
      "\n",
      "\n",
      "chemical\n",
      "air pollution\n",
      "\n",
      "\n",
      "psychosocial\n",
      "psychological well-being\n",
      "perceived stress\n",
      "depression\n",
      "reactivity to health stressors\n"
     ]
    }
   ],
   "source": [
    "#Outcome Layer1-Layer2 hierachy\n",
    "hierach3 = defaultdict(set)\n",
    "for i,v in enumerate(Outcome_Layer1):\n",
    "    hierach3[v].add(Outcome_Layer2[i])\n",
    "\n",
    "for k in hierach3.keys():\n",
    "    print('\\n')\n",
    "    print(k)\n",
    "    for v in hierach3[k]:\n",
    "        #print('\\t'+v)\n",
    "        print(v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "DNAmethylation\n",
      "genes-selected\n",
      "aging clocks\n",
      "genes-450K\n",
      "\n",
      "\n",
      "white blood cell counts\n",
      "basophil concentration\n",
      "neutrophil concentration\n",
      "eosinophil concentration\n",
      "total white blood cell\n",
      "monocytes concentration\n",
      "lymphocyte concentration\n",
      "\n",
      "\n",
      "blood pressure\n",
      "systolic blood pressure\n",
      "diastolic blood pressure\n",
      "\n",
      "\n",
      "cognitive function\n",
      "cognitive test z-score\n",
      "MMSE score\n",
      "global cognitive score\n",
      "verbal fluency\n",
      "immediate recall\n",
      "\n",
      "\n",
      "serum uric acid\n",
      "serum uric acid\n",
      "\n",
      "\n",
      "incident chronic kidney disease\n",
      "incident chronic kidney disease\n",
      "\n",
      "\n",
      "blood urea nitrogen\n",
      "blood urea nitrogen\n",
      "\n",
      "\n",
      "glomerular filtration rate\n",
      "glomerular filtration rate\n",
      "\n",
      "\n",
      "metabolites\n",
      "untargeted metabolomics profiling\n",
      "metabolites\n",
      "\n",
      "\n",
      "electrocardiogram\n",
      "heart rate variability markers\n",
      "JT interval\n",
      "QRS duration\n",
      "QT interval\n",
      "\n",
      "\n",
      "leukocyte telomere length\n",
      "leukocyte telomere length\n",
      "\n",
      "\n",
      "lung function\n",
      "FEV1/FVC\n",
      "FEF25%-75%\n",
      "FEV1\n",
      "FVC\n",
      "MMEF\n",
      "\n",
      "\n",
      "hemoglobin concentration\n",
      "hemoglobin concentration\n",
      "\n",
      "\n",
      "urinary metals\n",
      "urinary lead\n",
      "\n",
      "\n",
      "intraocular pressure\n",
      "intraocular pressure\n",
      "\n",
      "\n",
      "cardiovascular disease\n",
      "incident cardiovascular disease\n",
      "incident coronary heart disease\n",
      "acute coronary syndromes\n",
      "\n",
      "\n",
      "mortality\n",
      "mortality\n",
      "\n",
      "\n",
      "leukocyte distribution\n",
      "leukocyte distribution\n",
      "\n",
      "\n",
      "resistant hypertension\n",
      "resistant hypertension\n",
      "\n",
      "\n",
      "inflammation biomarkers\n",
      "inflammation biomarkers\n",
      "\n",
      "\n",
      "air pollution\n",
      "short-term PM2.5 elemental components\n",
      "short-term PM2.5 mass\n",
      "\n",
      "\n",
      "glaucoma\n",
      "incident primary open-angle glaucoma\n",
      "\n",
      "\n",
      "toenail metals\n",
      "toenail cadmium\n",
      "\n",
      "\n",
      "incident cancer\n",
      "incident cancer\n",
      "\n",
      "\n",
      "mitochondrial DNA\n",
      "mitochondrial DNA 8-OHdG lesions\n",
      "mitochondrial DNA copy number\n",
      "mitochondrial DNA lesions\n",
      "\n",
      "\n",
      "diabetes\n",
      "type2 diabetes\n",
      "\n",
      "\n",
      "bone lead\n",
      "tibia lead\n",
      "patella lead\n",
      "\n",
      "\n",
      "blood lead\n",
      "blood lead\n",
      "\n",
      "\n",
      "incident abnormal cardiac conductivity\n",
      "arrhythmia\n",
      "JT prolongation\n",
      "IVCD\n",
      "QT prolongation\n",
      "AVCD\n",
      "\n",
      "\n",
      "incident metabolic syndrome\n",
      "incident metabolic syndrome\n",
      "\n",
      "\n",
      "metabolic syndrome components\n",
      "metabolic syndrome components\n",
      "\n",
      "\n",
      "fasting blood glucose\n",
      "fasting blood glucose\n",
      "\n",
      "\n",
      "lipid profile\n",
      "lipid profile\n",
      "\n",
      "\n",
      "microRNA\n",
      "microRNA in extracellular vesicles\n",
      "microRNA in blood leukocytes\n",
      "\n",
      "\n",
      "reactivity to health stressors\n",
      "reactivity to health stressors\n",
      "\n",
      "\n",
      "psychological well-being\n",
      "eudaimonic well-being\n",
      "hedonic well-being\n",
      "\n",
      "\n",
      "olfactory recognition\n",
      "olfactory recognition\n",
      "\n",
      "\n",
      "depression\n",
      "depression\n",
      "\n",
      "\n",
      "tremor\n",
      "tremor\n",
      "\n",
      "\n",
      "perceived stress\n",
      "perceived stress\n",
      "\n",
      "\n",
      "plasma homocysteine\n",
      "plasma homocysteine\n",
      "\n",
      "\n",
      "leukocyte telomere length change rate\n",
      "leukocyte telomere length change rate\n",
      "\n",
      "\n",
      "arterial stiffness\n",
      "augmentation index\n",
      "augmentation pressure\n",
      "\n",
      "\n",
      "alzheimer disease biomarkers\n",
      "beta-amyloid\n",
      "\n",
      "\n",
      "intima-media thickness of the common carotid artery\n",
      "intima-media thickness of the common carotid artery\n",
      "\n",
      "\n",
      "neural plasticity\n",
      "acoustic startle reflex\n",
      "\n",
      "\n",
      "fine motor function\n",
      "grooved pegboard manual dexterity task\n",
      "handwriting tasks\n",
      "\n",
      "\n",
      "cholesterol level\n",
      "high-density lipoprotein cholesterol\n",
      "\n",
      "\n",
      "periodontal disease progression\n",
      "periodontal disease progression\n",
      "\n",
      "\n",
      "tooth loss\n",
      "tooth loss\n",
      "\n",
      "\n",
      "endothelial dysfunction markers\n",
      "endothelial dysfunction markers\n",
      "\n",
      "\n",
      "oxidative DNA damage\n",
      "oxidative DNA damage\n"
     ]
    }
   ],
   "source": [
    "#Outcome Layer2-Layer3 hierachy\n",
    "hierach4 = defaultdict(set)\n",
    "for i,v in enumerate(Outcome_Layer2):\n",
    "    hierach4[v].add(Outcome_Layer3[i])\n",
    "\n",
    "for k in hierach4.keys():\n",
    "    print('\\n')\n",
    "    print(k)\n",
    "    for v in hierach4[k]:\n",
    "        #print('\\t'+v)\n",
    "        print(v)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
